crates/runtime/src/dialogue/handler_safe_dialogue.rs

```diff
@@ -0,0 +1,209 @@
+use crate::prelude::*;
+use log::{debug, error};
+use std::ops::Deref;
+use yarn_slinger_core::prelude::*;
+
+/// A view of a [`Dialogue`]. Represents the subset of methods that are okay to be called from handlers.
+/// Since this type is `Send + Sync`, you can get a copy with [`Dialogue::get_handler_safe_dialogue`] and `move` it into a handler.
+///
+/// ## Implementation notes
+///
+/// This type is not present in the original. We need to use it to cleanly borrow data from handlers.
+/// The original just calls [`Dialogue`] for both mutable and immutable access anywhere,
+/// which is of course a big no-no in Rust.
+#[derive(Debug)]
+pub struct HandlerSafeDialogue {
+    pub(crate) log_debug_message: Logger,
+    pub(crate) log_error_message: Logger,
+    shared_state: SharedState,
+}
+
+impl SharedStateHolder for HandlerSafeDialogue {
+    fn shared_state(&self) -> &SharedState {
+        &self.shared_state
+    }
+}
+
+impl HandlerSafeDialogue {
+    pub(crate) fn from_shared_state(shared_state: SharedState) -> Self {
+        // Can't use a closure because the Rust type inference gets a bit confused :<
+        fn default_logger(msg: String, _dialogue: &HandlerSafeDialogue) {
```

Comment:
why does it need the _dialogue?

Reply:
Because the signature of the handler requires it. Long story. This is yeeted in #103